horovod/common/controller.cc

```diff
@@ -0,0 +1,735 @@
+// Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <atomic>
+#include <map>
+#include <queue>
+#include <set>
+#include <unordered_set>
+
+#include "controller.h"
+#include "global_state.h"
+#include "logging.h"
+#include "operations.h"
+
+namespace horovod {
+namespace common {
+
+Controller::Controller(HorovodGlobalState& global_state)
+    : tensor_queue_(global_state.tenor_queue),
+      timeline_enabled_(global_state.timeline_enabled),
+      timeline_(global_state.timeline),
+      response_cache_(global_state.response_cache),
+      parameter_manager_(global_state.parameter_manager),
+      shut_down_(global_state.shut_down),
```

Comment:
Shutdown flag doesn't feel right as a dependency-injected parameter.  I suggest switching shutdown flag to be passed to ComputeResponseList()